src/compute/sort/row/mod.rs

```diff
@@ -0,0 +1,1041 @@
+
+                let mapping = compute_dictionary_mapping(interner, values)?
+                    .into_iter()
+                    .map(|maybe_interned| {
+                        maybe_interned.map(|interned| interner.normalized_key(interned))
+                    })
+                    .collect::<Vec<_>>();
+
+                Ok(Some(mapping))
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        let mut rows = new_empty_rows(columns, &dictionaries, Arc::clone(&self.fields))?;
+
+        for ((column, field), dictionary) in
+            columns.iter().zip(self.fields.iter()).zip(dictionaries)
+        {
+            // We encode a column at a time to minimise dispatch overheads
+            encode_column(&mut rows, column, field.options, dictionary.as_deref())
+        }
+
+        Ok(rows)
+    }
+}
+
+/// A row-oriented representation of arrow data, that is normalized for comparison
+///
+/// See [`RowConverter`]
+#[derive(Debug)]
+pub struct Rows {
+    /// Underlying row bytes
+    buffer: Box<[u8]>,
+    /// Row `i` has data `&buffer[offsets[i]..offsets[i+1]]`
+    offsets: Box<[usize]>,
+    /// The schema for these rows
+    fields: Arc<[SortField]>,
+}
+
+impl Rows {
+    /// Get a reference to a certain row.
+    pub fn row(&self, row: usize) -> Row<'_> {
+        let end = self.offsets[row + 1];
+        let start = self.offsets[row];
+        Row {
+            data: &self.buffer[start..end],
+            _fields: &self.fields,
+        }
+    }
+
+    /// Returns the number of rows
+    pub fn num_rows(&self) -> usize {
+        self.offsets.len() - 1
+    }
+
+    /// Returns the iterator
+    pub fn iter(&self) -> RowsIter<'_> {
+        self.into_iter()
+    }
+}
+
+impl<'a> IntoIterator for &'a Rows {
+    type Item = Row<'a>;
+    type IntoIter = RowsIter<'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        RowsIter {
+            rows: self,
+            start: 0,
+            end: self.num_rows(),
+        }
+    }
+}
+
+/// An iterator over [`Rows`]
+#[derive(Debug)]
+pub struct RowsIter<'a> {
+    rows: &'a Rows,
+    start: usize,
+    end: usize,
+}
+
+impl<'a> Iterator for RowsIter<'a> {
+    type Item = Row<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.end == self.start {
+            return None;
+        }
+        let row = self.rows.row(self.start);
+        self.start += 1;
+        Some(row)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = self.len();
+        (len, Some(len))
+    }
+}
+
+impl<'a> ExactSizeIterator for RowsIter<'a> {
+    fn len(&self) -> usize {
+        self.end - self.start
+    }
+}
+
+impl<'a> DoubleEndedIterator for RowsIter<'a> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        if self.end == self.start {
+            return None;
+        }
+        let row = self.rows.row(self.end);
+        self.end -= 1;
+        Some(row)
+    }
+}
+
```

Comment:
```suggestion

unsafe impl<'a> crate::trusted_len::TrustedLen for RowsIter<'a> {}
```

since this iterator is trusted len.